blog: Add GSoC'24 blog on debugging mimalloc port (#448)
unikraft/docs: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx

---
title: "GSoC'24: Debugging the Multithreaded Memory Allocation Interface"
description: |
  Different applications have different memory usage patterns and perform better
  when using dedicated memory allocators. Right now, the buddy allocator is the only
  available memory allocator used in Unikraft, so this GSoC project aims to 
  port more memory allocators to it, starting with mimalloc.
publishedDate: 2024-08-01
image: /images/unikraft-gsoc24.png
authors:
- Yang Hu
tags:
- gsoc
- gsoc24
- allocators
- synchronization
---

## Project Overview

### Memory allocation in Unikraft

Memory allocators have a significant impact on application performance.

> [Research](https://dl.acm.org/doi/10.1145/378795.378821) have shown that programs can run as much as 60% faster just by switching to an appropriate memory allocator.

Reviewer: There have been some [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators. According to their work, Switching to the appropiate memory allocator may improve the application performance by 60%.

Unikraft currently supports only one general-purpose memory allocator, the buddy allocator.
This project aims to port [mimalloc](https://github.com/microsoft/mimalloc) (pronounced "me-malloc"), a high-performance general-purpose memory allocator developed by Microsoft, to Unikraft.

> This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to optimize the performance of their applications.

Reviewer: ```suggestion
This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to enhance the performance of their applications.
```

### Objectives

> [Hugo Lefeuvre](https://github.com/hlef) has made an effort to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).

Reviewer: ```suggestion
[Hugo Lefeuvre](https://github.com/hlef) made an initial work to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
```

> However, as Unikraft has evolved significantly over the years, more work is needed to adapt the `lib-mimalloc` port to the latest Unikraft core.

Reviewer: ```suggestion
However, as Unikraft has evolved significantly over the years, more work has to be done in order to adapt the `lib-mimalloc` port to the latest Unikraft core.
```

So far, I have successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc `v1.6.3` against the latest Unikraft core (v0.17.0) with `lib-musl`.

> Now, I am focused on validating the port and testing its performace.

Reviewer: ```suggestion
Now, I am focused on validating the functionality and benchmarking the performance of the allocator.
```

> After that, the goal is to port the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0).

Reviewer: ```suggestion
The next step would be to port the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0).
```

## Current Progress

> Following porting the multithreaded `cache-scratch` stress test in [the last blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft), this past three weeks I have been debugging the it.

Reviewer: ```suggestion
Following porting the multithreaded `cache-scratch` stress test in [the last blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft), this past three weeks I have been debugging the it.
```

This is unclear to me, can you rephrase?

Author: Done

### Debugging `pthread`

The [first error we ran into](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft#testing) was a memory corruption error with the `pthread` interface.

> After investigating the `pthread_join()` calls, I found that the threads were never even created correctly at all because `pthread_create()` uses the POSIX `mmap` system call, which never existed as we did not turn on virtual memory support (i.e., `ukvmem` alongside `posix-mmap`).

Reviewer: I don't think you should go that deep into technical details

Author: I wrote the blog in sort of chronological order - it was because of this`pthread` bug that I turned on virtual memory support, hence the rest. I have shortened the paragraph but still kept it here.

[   11.476011] CRIT: <idle> {r:0x10709c,f:0x28cf00} [libkvmplat] <trace.c @   50> R10: 0000000000000000 R11: 000000000000002d R12: 0000000000000000
[   11.493210] CRIT: <idle> {r:0x1070e8,f:0x28cf00} [libkvmplat] <trace.c @   52> R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[   11.510388] CRIT: <idle> {r:0x107240,f:0x28cf00} [libkvmplat] <trace.c @   86> base is 0x11a70 caller is 0x1732e6
```

After extensive debugging, I realized that this bug is not unique to the *mimalloc* port.
I reproduced a similar error with the same library configurations on the buddy allocator as well and I have submitted [an issue](https://github.com/unikraft/unikraft/issues/1475) addressing this.
More work needs to be done to fix the virtual memory infrastructure.

## Next steps

In order to resolve the abovementioned issues, I still have much to study about Unikraft's virtual memory management subsystem, its boot process, and how *mimalloc* works under the hood - and I will focus on just that in the following weeks.

Hopefully, with those issues fixed, we could finally get the `cache-scratch` test running on multiple cores.

> Once that is done, I will get back on track with my "global" TODO list:

Reviewer: Can you mention what ideas do you have for benchmarking?

Author: I have written much about benchmarking (i.e., which benchmarks I would use, what they are for, etc.) in my last blog post so I didn't elaborate here. I will attach a link to my last blog post

---
title: "GSoC'24: Debugging the Multithreaded Memory Allocation Interface"
description: |
  Different applications have different memory usage patterns and perform better
  when using dedicated memory allocators. Right now, the buddy allocator is the only
  available memory allocator used in Unikraft, so this GSoC project aims to 
  port more memory allocators to it, starting with mimalloc.

> publishedDate: 2024-07-12

Reviewer 2: Change the date

  * But then, we would have to cope with VMA objects from different
  * allocators. So instead, we just initialize a small heap using fixed
  * mappings and then create the VMA on top. Afterwards, we add the
  * remainder of the VMA to the allocator.
  */
```

This is because Unikraft's Virtual Memory Areas (VMAs, described by `struct uk_vma`) are managed by a Virtual Address Space (VAS, described by `struct uk_vas`), which has one dedicated memory allocator assigned to it.
So it would be impractical to allocate two VMAs under a same VAS using two different memory allocators.

The code for this logic though, is customized for the buddy allocator:

1. Initialize a physically-mapped small heap size of 16 pages
1. Initialize the allocator based on this heap
1. Initialize the VAS
1. Map the rest of the heap as an anonymous VMA
1. Call `uk_alloc_addmem()` to let the buddy allocator know that this anonymous region that is at its disposal

This design accommodates the fact that the buddy allocator needs initial memory space to store its metadata and that it needs information about the total available memory for initializing its free lists.

> However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs are not yet ready.

Reviewer 2: ```suggestion
However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs is not yet ready.
```

[   11.441611] CRIT: <idle> {r:0x107004,f:0x28cf00} [libkvmplat] <trace.c @   46> RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
[   11.458821] CRIT: <idle> {r:0x107050,f:0x28cf00} [libkvmplat] <trace.c @   48> RBP: 0000000000011a70 R08: 0000000000000000 R09: 0000000000000000
[   11.476011] CRIT: <idle> {r:0x10709c,f:0x28cf00} [libkvmplat] <trace.c @   50> R10: 0000000000000000 R11: 000000000000002d R12: 0000000000000000
[   11.493210] CRIT: <idle> {r:0x1070e8,f:0x28cf00} [libkvmplat] <trace.c @   52> R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[   11.510388] CRIT: <idle> {r:0x107240,f:0x28cf00} [libkvmplat] <trace.c @   86> base is 0x11a70 caller is 0x1732e6
```

After extensive debugging, I realized that this bug is not unique to the *mimalloc* port.
I reproduced a similar error with the same library configurations on the buddy allocator as well and I have submitted [an issue](https://github.com/unikraft/unikraft/issues/1475) addressing this.

The current `ukvmem` implementation is also incomplete as some of its key data structures are not CPU-local.
For example,

> At `vmem.c:27`:

Reviewer 2: Link the file on github.

Author: @mogasergiu has pointed out that this should not be a concern at this time so this part has been removed

```c
/*
 * Pointer to currently active virtual address space.
 * TODO: This should move to a CPU-local variable
 */
static struct uk_vas *vmem_active_vas;
```

> And `paging.c:97`:

Reviewer 2: Link the file
